### PR TITLE
feat: migrate EVM interpreter to custom U256

### PR DESCRIPTION
**Pull request by 0xVolosnikov**
## What ❔

An attempt to revive https://github.com/matter-labs/zksync-os/pull/135

This PR migrates the EVM interpreter-facing path from `ruint::aliases::U256` to the custom non-`Copy` `u256::U256`.

It is split into three reviewable commits:
- shared `zk_ee` helpers and trait-bound updates for the new integer type
- `evm_interpreter` migration of stack, frame state, and opcode implementations
- `forward_system` tracer/logger adaptation at the EVM API boundary

## Why ❔

The proving-oriented custom `U256` needs to be exercised in the EVM execution path without forcing a workspace-wide integer migration in one step.

This keeps the change localized to the interpreter and its tracing boundary, reduces copy-heavy assumptions in the EVM stack/frame code, and makes the migration reviewable in commit-sized pieces.

## Is this a breaking change?
- [ ] Yes
- [x] No

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted.

### CODE REVIEW

**Comment by 0xVolosnikov** [comment]
Notes for later

To research: maybe reusing old EVM frames (including stack) would give better results

TODO: add fuzzing for 2 u256 implementations to check consistency
TODO: propagate new custom u256 further (to not have several different implementations used)

**Comment by 0xVolosnikov** [comment]
Huh, it improves benchmarks **_without_** delegation enabled: 

### Benchmark report

| Benchmark | Symbol | Base Eff | Head Eff (%) | Base Raw | Head Raw (%) | Base Blake | Head Blake (%) | Base Bigint | Head Bigint (%) |
|-----------|--------|-----------|----------------|-----------|----------------|-------------|------------------|---------------|--------------------|
| `block_19299001` | `process_block` | 299,079,274 | 295,204,533 (-1.30%) | 263,748,186 | 259,873,445 (-1.47%) | 410,630 | 410,630 (+0.00%) | 7,190,252 | 7,190,252 (+0.00%)
| `block_22244135` | `process_block` | 183,557,041 | 178,672,860 (-2.66%) | 162,838,557 | 157,954,376 (-3.00%) | 172,040 | 172,040 (+0.00%) | 4,491,461 | 4,491,461 (+0.00%)

**Comment by 0xVolosnikov** [comment]
@vv-dev-ai ptal

**Comment by vv-dev-ai** [comment]
[AI-review] Reviewed commit `1963b5396a8e7f3fd5c044527fcdb4fdc1302d9e` — **COMMENT** (2 inline comments).

File: evm_interpreter/src/instructions/bitwise.rs [hunk 1]
```diff
@@ -145,8 +157,30 @@ impl<S: EthereumLikeTypes> Interpreter<'_, S> {
...
+            if is_negative {
+                // Sign-extend: OR with a mask of 1s in the top `shift` bits
+                let mut mask = U256::zero();
+                mask.not_mut(); // all 1s
+                mask <<= (256 - shift) as u32;
+                core::ops::BitOrAssign::bitor_assign(op2, &mask);
+            }
+        }
         Ok(())
     }
```
Comment (vv-dev-ai) [inline evm_interpreter/src/instructions/bitwise.rs]: [AI-review] **Low:** Manual SAR replaces `arithmetic_shr` — logic verified correct, but this is a non-trivial reimplementation.
Trigger: Boundary shift values (0, 255, 256) combined with negative inputs like `MIN_INT` (`0x8000...0000`).
Impact: Incorrect arithmetic right shift would break signed-division-dependent Solidity patterns.
Fix: Add unit tests covering SAR with shifts {0, 1, 128, 255, 256, 257} × {positive, negative, MIN_INT, -1, 0}.

File: evm_interpreter/src/evm_stack.rs [hunk 2]
```diff
@@ -295,14 +298,53 @@ impl<A: Allocator> EvmStack<A> {
         }
     }
 
+    /// Pop 2 elements as mutable and peek the next one as mutable.
+    /// Returns ((op1_mut, op2_mut), peek_mut) where op1 was on top.
+    /// After the call, the stack length is reduced by 2 (the peeked element stays).
+    #[inline(always)]
+    pub fn pop_2_mut_and_peek(
+        &'_ mut self,
+    ) -> Result<((&'_ mut U256, &'_ mut U256), &'_ mut U256), ExitCode> {
```
Comment (vv-dev-ai) [inline evm_interpreter/src/evm_stack.rs:298]: [AI-review] **Nit:** New `unsafe` block follows the existing pattern in this file but lacks a `// SAFETY:` comment. The three `&mut U256` references don't alias because they point to indices `len-1`, `len-2`, `len-3` respectively (guaranteed distinct by the `self.len < 3` guard). Worth documenting for the same reason the other pop functions should have it.

**Review by vv-dev-ai — COMMENTED** [review]
[AI-review] PR #588 — **COMMENT**

Migration from `ruint::aliases::U256` to `u256::U256` across the EVM interpreter path is correct. Verified semantic equivalence for all affected opcodes, i256 signed operations, stack operations, and the forward-system tracing boundary.

- In-place `div_rem`/`add_mod`/`mul_mod` semantics correctly mapped in DIV, MOD, SDIV, SMOD, ADDMOD, MULMOD
- Clone-based swap in `EvmStack::swap` equivalent to old `mem::swap` for the non-Copy type
- Forward/proving boundary clean: `delegation` feature propagates correctly through the crate tree, `u256::init()` placed correctly in proving setup
- `write_into_ptr` signature unified between naive and RISC-V implementations (good fix)
- Benchmark evidence (1.3–3% effective cycle reduction) present in PR comments

### Suggested Tests
- **SAR (arithmetic right shift):** Rewritten from `arithmetic_shr` to manual shift + sign-extension mask. Test boundary shifts (0, 1, 128, 255, 256, >256) for both positive and negative inputs, especially `MIN_INT` (`0x8000...0000`) and `-1`.
- **ADDMOD/MULMOD:** Verify the in-place `U256::add_mod`/`mul_mod` with modulus=0, modulus=1, and operands larger than modulus through the new `pop_2_mut_and_peek` path.

File: evm_interpreter/src/evm_stack.rs (hunk 2 above)
Comment (0xVolosnikov) [reply evm_interpreter/src/evm_stack.rs:298]: Added in c1d56c68. The unsafe block now documents why the three returned mutable references are derived from distinct stack slots and therefore do not alias.

Comment (0xVolosnikov) [reply evm_interpreter/src/evm_stack.rs:298]: Added in c1d56c68. The unsafe block now documents why the three returned mutable references are derived from distinct stack slots and therefore do not alias.

File: evm_interpreter/src/instructions/bitwise.rs (hunk 1 above)
Comment (0xVolosnikov) [reply evm_interpreter/src/instructions/bitwise.rs]: Added in c1d56c68. There is now a boundary test that checks SAR against the host `ruint` reference for shifts `{0,1,128,255,256,257}` across positive, negative, `MIN_INT`, `-1`, and `0` cases.

File: evm_interpreter/src/evm_stack.rs [hunk 3]
```diff
@@ -69,18 +69,21 @@ impl<A: Allocator> EvmStack<A> {
...
+                .assume_init_mut();
+            let dst = self
+                .buffer
+                .as_mut_ptr()
+                .add(dst_offset)
+                .as_mut_unchecked()
+                .assume_init_mut();
+            let t = src.clone();
+            Clone::clone_from(src, &*dst);
+            Clone::clone_from(dst, &t);
```
[the remark this answers is not in the document]
Comment (0xVolosnikov) [reply evm_interpreter/src/evm_stack.rs]: Addressed in c1d56c68. `swap()` now uses in-place `core::mem::swap` on initialized `&mut U256` slots, so the delegated backend no longer pays the extra clone cost here.

File: forward_system/src/run/tracing_impl.rs [hunk 4]
```diff
@@ -26,23 +26,36 @@ struct ExecutionEnvironmentLaunchParamsWrapped<'a, 'b, S: EthereumLikeTypes>(
...
 impl<'a, S: EthereumLikeTypes + 'a, T: EvmFrameInterface<S>> From<&'a T>
     for EvmFrameInterfaceWrapped<'a, S, T>
 {
     fn from(value: &'a T) -> Self {
+        let call_value_converted: ruint::aliases::U256 = value.call_value().clone().into();
+        let stack_values = value
+            .stack()
+            .to_slice()
+            .iter()
+            .map(|value| -> U256 { value.clone().into() })
```
[the remark this answers is not in the document]
Comment (0xVolosnikov) [reply forward_system/src/run/tracing_impl.rs]: Addressed in c1d56c68. The tracer bridge now converts the stack lazily via `OnceCell`, so we avoid the eager per-opcode `Vec` clone and only materialize converted values when a tracer actually inspects the stack.

**Comment by 0xVolosnikov** [comment]
@shamatar initial re-integration of custom U256 in EVM interpreter. Maybe you'll have some additional comments or ideas

**Comment by shamatar** [comment]
Can you trigger performance run on it?

**Comment by shamatar** [comment]
Also, do you want to add division oracles in the same PR or in the separate one?

**Comment by 0xVolosnikov** [comment]
> Can you trigger performance run on it?

Do you mean benchmarks? If this is the case - they are posted above. 

> Also, do you want to add division oracles in the same PR or in the separate one?

I think it's fine to do it in the same PR. The idea here is to reuse the same queries used by e.g. modexp?

**Comment by vv-dev-ai** [comment]
[AI-review] Reviewed commit `d1356b4c5d13630434a8116583f79a41c257c9cd` — **APPROVE** (0 inline comments).

**Review by vv-dev-ai — APPROVED** [review]
[AI-review] PR #588 — **APPROVE**

- ✅ RESOLVED `sar-manual-impl`: SAR boundary tests added — verified: 30 test cases (shifts {0,1,128,255,256,257} × {positive, negative, MIN_INT, -1, 0}) validated against `ruint` reference.
- ✅ RESOLVED (Nit): `// SAFETY:` comment added to `pop_2_mut_and_peek` documenting non-aliasing of the three mutable references.

**Comment by 0xVolosnikov** [comment]
TODO: mulmod

File: callable_oracles/src/arithmetic/mod.rs [hunk 5]
```diff
@@ -19,9 +24,50 @@ impl<M: MemorySource> Default for ArithmeticQuery<M> {
...
+    /// Handle U256 div_rem oracle query.
+    ///
+    /// Input: 1 packed usize containing two u32 pointers (dividend_ptr in low, divisor_ptr in high).
+    /// Output: 8 usize values — 4 u64 limbs for quotient, then 4 u64 limbs for remainder (LE).
+    fn process_u256_div_rem_query(
+        &self,
+        query: Vec<usize>,
+        memory: &M,
+    ) -> Box<dyn ExactSizeIterator<Item = usize> + 'static + Send + Sync> {
+        let mut it = query.into_iter();
```
Comment (shamatar) [inline callable_oracles/src/arithmetic/mod.rs]: I would instead use a "structure" approach as for modexp, so query is just a single pointer, and the structure provides pointers to all the needed inputs

File: supporting_crates/u256/src/risc_v/mod.rs [hunk 6]
```diff
@@ -3,6 +3,126 @@ use core::ops::{
...
+/// Oracle query ID for U256 division hints.
+/// Must match `zk_ee::oracle::query_ids::U256_DIV_REM_ADVICE_QUERY_ID`.
+#[cfg(target_arch = "riscv32")]
+const U256_DIV_REM_ADVICE_QUERY_ID: u32 = 0x4005_0030;
+
+/// Write a word to the oracle CSR (address 0x7c0).
+/// Mirrors `riscv_common::csr_write_word`.
+#[cfg(target_arch = "riscv32")]
+#[inline(always)]
+fn oracle_csr_write(value: usize) {
```
Comment (shamatar) [inline supporting_crates/u256/src/risc_v/mod.rs:21]: Don't we provide this in riscv_common crate in airbender itself?

**Comment by shamatar** [comment]
From the stats - looks like we want custom conversions from u8/u64

**Comment by 0xVolosnikov** [comment]
Also a note: current per-opcode benchmarks do not account for bigint delegations

**Comment by shamatar** [comment]
Sure, we will need to count delegation cycles with a proper cost-equivalent coefficient, but there is hardly a fundamental reason for conversion from primitive types to become more expensive with custom U256

**Comment by 0xVolosnikov** [comment]
Yep, will try to fix it. Also AI provided a bunch of other observations/ideas, will iterate over them

File: evm_interpreter/src/evm_stack.rs [hunk 7]
```diff
@@ -360,6 +360,43 @@ impl<A: Allocator> EvmStack<A> {
...
+    /// Push a u64 value as U256 directly into the next stack slot.
+    /// Uses a single delegation (zero + limb write) instead of the
+    /// two-delegation path of U256::from(u64) + push.
+    #[inline(always)]
+    pub fn push_u64(&mut self, value: u64) -> Result<(), ExitCode> {
+        if self.len == STACK_SIZE {
+            return Err(EvmError::StackOverflow.into());
+        }
+        unsafe {
+            let dst = self.buffer.as_mut_ptr().add(self.len).as_mut_unchecked();
```
Comment (shmatar) [inline evm_interpreter/src/evm_stack.rs:372]: Even though I wrote it, it’s unsound. We should write into pointer, as we do now know if memory location was initialized before or not (as we push and not pop/push). Some other ops may need similar fixes

Comment (shmatar) [reply evm_interpreter/src/evm_stack.rs:372]: Here just removing as_mut_unchecked would be enough, as we go back into pointer anyway below

File: supporting_crates/u256/src/risc_v/mod.rs [hunk 8]
```diff
@@ -38,20 +158,42 @@ impl U256 {
...
     }
 
-    pub unsafe fn write_into_ptr_unchecked(&self, dst: *mut Self) {
-        delegated_u256::write_into_ptr_unchecked(dst.cast(), &self.0);
+    /// # Safety
+    /// `a` and `b` must be valid, properly aligned pointers to initialized `Self` values.
+    ///
+    /// On the delegated backend this is cheaper than a generic `mem::swap`, because it stays on
+    /// the bigint memcopy path instead of forcing a raw 32-byte move sequence in RISC-V code.
+    pub unsafe fn swap_in_place(a: *mut Self, b: *mut Self) {
```
Comment (shamatar) [inline supporting_crates/u256/src/risc_v/mod.rs:182]: What about making all swaps non-overlapping (and rename function signature)? We do not have any other cases in practice

Comment (shamatar) [reply supporting_crates/u256/src/risc_v/mod.rs:182]: Creating a local uninit value may require stack alignment, so it may be better to put inline-always here, so everything is embedded into the caller

File: supporting_crates/u256/src/risc_v/mod.rs [hunk 9]
```diff
@@ -38,20 +158,42 @@ impl U256 {
...
+            Self::write_into_ptr_unchecked(b, tmp.assume_init_ref());
+        }
     }
 
     pub fn clone_into(&self, dst: &mut Self) {
-        unsafe { self.write_into_ptr(dst as *mut _) };
+        unsafe { Self::write_into_ptr(dst as *mut _, self) };
     }
 
     pub unsafe fn clone_into_unchecked(&self, dst: &mut Self) {
```
Comment (shamatar) [inline supporting_crates/u256/src/risc_v/mod.rs:199]: such short functions are usually inlined, but I usually prefer to make it explicit

**Comment by vv-dev-ai** [comment]
[AI-review] Reviewed commit `d8589835ecf6a6fffc00b7348645edb8f68cc8bb` — **COMMENT** (0 inline comments).

**Review by vv-dev-ai — COMMENTED** [review]
[AI-review] PR #588 — **COMMENT**

Reply-review: evaluating shamatar's feedback. All points are valid design suggestions, no correctness blockers.

**On shamatar's comments:**
- **Oracle query structure** (`callable_oracles/src/arithmetic/mod.rs`): Agree — a `repr(C)` struct (like `ModExpAdviceParams`) would be more readable and extensible than the packed-pointer approach.
- **CSR helper duplication** (`supporting_crates/u256/src/risc_v/mod.rs`): Agree — functions explicitly mirror `riscv_common` utilities. Import from `riscv_common` if the dependency graph allows.
- **`swap_in_place` `#[inline(always)]`** (line 178): Agree — the `MaybeUninit` temporary may force stack frame allocation in RISC-V; inlining avoids this.
- **`clone_into_unchecked` inline** (line 195): Agree — single-line wrapper should match other `#[inline(always)]` utilities in this file.

**New `oracle_div_rem` correctness** (`supporting_crates/u256/src/risc_v/mod.rs:49-124`): Verification logic is sound — `q * d + r == n` via widening multiplication, overflow check (high == 0, no carry), and `r < d` uniquely determine quotient and remainder. Full `assert_eq!` on response length correctly prevents CSR stream corruption.

**Note:** 8 new commits since last review add oracle-backed div_rem, `push_u64`/`push_b160` stack optimizations, and PUSH1 specialization. A full triggered review is recommended for comprehensive coverage.
